### arange

在給定的間隔內返回均勻間隔的值。[[參考資料](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.arange.html)]

In [1]:
import numpy as np

np.arange(0,10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [2]:
np.arange(0,10,2)

array([0, 2, 4, 6, 8])

### zeros and ones

In [3]:
np.zeros(3)

array([0., 0., 0.])

In [6]:
np.zeros((5,5))

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [7]:
np.ones(3)

array([1., 1., 1.])

In [8]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

### linspace 

In [9]:
np.linspace(0,10,3)

array([ 0.,  5., 10.])

In [10]:
np.linspace(0,1,10)

array([0.        , 0.11111111, 0.22222222, 0.33333333, 0.44444444,
       0.55555556, 0.66666667, 0.77777778, 0.88888889, 1.        ])

### eye

In [11]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

## Random 
NumPy也有許多方法來創建隨機數陣列：

In [20]:
np.random.rand(2)

array([0.28508846, 0.71565445])

In [27]:
np.random.rand(5,2)

array([[0.00765819, 0.32850452],
       [0.58152646, 0.36451677],
       [0.80487157, 0.61885883],
       [0.63862843, 0.40044316],
       [0.57570847, 0.20961472]])

In [44]:
V = np.random.rand(100)
m = V.mean()
std = V.std()
m, std

(0.5280765122899314, 0.28556822072225396)

### randn

從“標準正態”分佈[σ = 1]返回樣本（或樣本）。與 **rand** 不同，靠近零的值更有可能出現。[[參考資料](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.randn.html)]


標準正態分佈是一種特殊的常態分佈，其平均數（μ）為0，標準差（σ）為1。在標準正態分佈中，數據點主要集中在平均值附近，並且隨著距離平均值的增加而呈指數下降。這種分佈的曲線形狀呈鐘形，並且在平均值附近有一個尖峰。

在機器學習和統計學中，標準正態分佈通常用於建模和分析，並且在許多算法和技術中被廣泛應用。

In [57]:
v = np.random.randn(100)
m = v.mean()
std = v.std()
m, std

(0.00854150145907177, 1.0252189000379524)

### randint

In [66]:
np.random.randint(1,100,10)

array([91, 67, 35, 27, 77, 87, 88, 37, 37,  7])

In [79]:
#不重複抽
np.random.choice(range(1,100),10,replace=False)

array([1, 3, 5])

### seed
可以用來設置隨機狀態，以便可以重現相同的“隨機”結果。[[參考資料](https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html)]

In [98]:
np.random.seed(42)
np.random.randn(4)

array([ 0.49671415, -0.1382643 ,  0.64768854,  1.52302986])

## Reshape
返回一個包含相同數據但具有新形狀的陣列。[[參考資料](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html)]

In [110]:
arr = np.arange(25)
ranarr = np.random.randint(0,50,10)
arr,ranarr

(array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
        17, 18, 19, 20, 21, 22, 23, 24]),
 array([13, 49,  8, 25,  1, 19, 27, 46,  6, 43]))

In [111]:
arr5x5 = arr.reshape(5,5)
print(arr)
print(arr5x5)
arr5x5[0,2] = 99

print(arr)
print(arr5x5)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24]
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]
[ 0  1 99  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24]
[[ 0  1 99  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


In [114]:
arr = np.arange(25)
newArr =  np.resize(arr,(5,6)) #資料會複製一份
print(arr)
print(newArr)
newArr[2,2] = 88

print(arr)
print(newArr)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24]
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]
 [24  0  1  2  3  4]]
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24]
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 88 15 16 17]
 [18 19 20 21 22 23]
 [24  0  1  2  3  4]]


### max, min, argmax, argmin

In [116]:
ranarr = np.random.randint(0,50,10)
ranarr

array([ 5, 41,  3, 28, 17, 25, 43, 33,  9, 35])

In [117]:
ranarr.max()

43

In [118]:
ranarr.argmax()

6

In [119]:
ranarr.min()


3

In [120]:
ranarr.argmin()

2

## Indexing-and-Selection

In [121]:
arr = np.arange(0,11)
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [122]:
arr[8]

8

In [123]:
arr[1:5]

array([1, 2, 3, 4])

In [124]:
arr[:5]

array([0, 1, 2, 3, 4])

In [125]:
arr[:5] = 100
arr

array([100, 100, 100, 100, 100,   5,   6,   7,   8,   9,  10])

In [126]:
arr = np.arange(0,11)
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [128]:
slice_of_arr = arr[0:7]
slice_of_arr
slice_of_arr[3] = 99

print(arr)
print(slice_of_arr)

[ 0  1  2 99  4  5  6  7  8  9 10]
[ 0  1  2 99  4  5  6]


In [129]:
myList = [1,2,3,4,5,6]
slice_list = myList[0:4]
slice_list[2] = 99
print(myList)
print(slice_list)

[1, 2, 3, 4, 5, 6]
[1, 2, 99, 4]


## 索引二維陣列（矩陣）

一般格式是 **arr_2d[row][col]** 或 **arr_2d[row, col]**。我推薦使用逗號標記法以增加清晰度。

In [131]:
arra_2d = np.array(([5,10,15],
                    [20,25,30],
                    [19,18,45],
                    [26,29,32]) ) 
arra_2d

array([[ 5, 10, 15],
       [20, 25, 30],
       [19, 18, 45],
       [26, 29, 32]])

In [132]:
arra_2d[1]

array([20, 25, 30])

In [133]:
arra_2d[1][0]

20

In [135]:
arra_2d[1,0]

20

In [136]:
arra_2d[:2,1:]

array([[10, 15],
       [25, 30]])

In [137]:
arra_2d[2,:]

array([19, 18, 45])

## 條件選擇


In [138]:
arr = np.arange(1,11)
arr

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [140]:
bool_arr = arr > 4
arr[bool_arr]

array([ 5,  6,  7,  8,  9, 10])

In [141]:
arr[arr > 2]

array([ 3,  4,  5,  6,  7,  8,  9, 10])

# NumPy 運算

In [145]:
arr =  np.arange(0,10)
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [147]:
print(arr + arr)
print(arr - arr)
print(arr * arr)

[ 0  2  4  6  8 10 12 14 16 18]
[0 0 0 0 0 0 0 0 0 0]
[ 0  1  4  9 16 25 36 49 64 81]


In [148]:
print(arr / arr)

[nan  1.  1.  1.  1.  1.  1.  1.  1.  1.]


C:\Users\xvpow\AppData\Local\Temp\ipykernel_8964\3831679784.py:1: RuntimeWarning: invalid value encountered in divide
  print(arr / arr)


In [149]:
np.sqrt(arr)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ])

In [150]:
# Calculating exponential (e^)
np.exp(arr)

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03])

In [151]:
np.sin(arr)

array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
       -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849])

In [152]:
np.cos(arr)

array([ 1.        ,  0.54030231, -0.41614684, -0.9899925 , -0.65364362,
        0.28366219,  0.96017029,  0.75390225, -0.14550003, -0.91113026])

在 NumPy 中使用 `np.log(arr)` 的意思是對 `arr` 這個陣列中的每個元素進行自然對數運算。這個函數計算的是以 e (約等於 2.71828) 為底的對數。自然對數在數學、物理學、工程學及經濟學等許多領域中都非常重要，常用於各種成長或衰減模型的分析。使用此函數時，必須確保陣列中的所有元素都是正數，因為對數只對正數有定義。

In [153]:
np.log(arr)

C:\Users\xvpow\AppData\Local\Temp\ipykernel_8964\3120950136.py:1: RuntimeWarning: divide by zero encountered in log
  np.log(arr)


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458])

## 陣列的摘要統計

NumPy 也提供了常見的摘要統計功能，如 <em>sum</em>（總和）、<em>mean</em>（平均值）和 <em>max</em>（最大值）。你可以將這些函數作為陣列的方法來調用。下面是如何使用這些統計方法的例子：

1. **np.sum(arr)** - 計算陣列 `arr` 中所有元素的總和。
2. **np.mean(arr)** - 計算陣列 `arr` 中所有元素的平均值。
3. **np.max(arr)** - 找出陣列 `arr` 中的最大元素。
4. **np.min(arr)** - 找出陣列 `arr` 中的最小元素。
5. **np.std(arr)** - 計算陣列 `arr` 中元素的標準差。
6. **np.var(arr)** - 計算陣列 `arr` 中元素的方差。

In [ ]:
arr =  np.arange(0,10)
arr


In [154]:
arr.sum()

45

In [155]:
arr.mean()

4.5

In [156]:
arr.max()

9

In [157]:
print(arr.var())
print(arr.std())#求標準差

8.25
2.8722813232690143
